safety-architecture/tools/callgraph-tool/scripts/find_callchains.py

```diff
@@ -0,0 +1,308 @@
+
+    # If there are no functions matching regex exit with info message
+    df_to = def_regex_filter(df, column=function_col, regex=to_fun)
+    df_to = df_to[[function_col, filename_col]].drop_duplicates()
+
+    if df_to.shape[0] <= 0:
+        _LOGGER.warn(
+            "Function regex '%s' does not match any entries in call graph database"
+            " for selected search direction" % to_fun
+        )
+        sys.exit(1)
+
+    return df_to
+
+
+def graph_from_df(df):
+    g = nx.DiGraph()
+    for row in df.itertuples():
+        n1 = Node(
+                function=row.caller_function,
+                filename=row.caller_filename)
+        n2 = Node(
+                function=row.callee_function,
+                filename=row.callee_filename)
+        g.add_edge(n1, n2)
+
+    return g
+
+
+def find_chains_directed_df(df, from_fun, to_fun, dir):
+    df_from = get_df_from(df, from_fun, dir.from_col_func, dir.from_col_fn)
+    df_to = get_df_to(df, to_fun, dir.to_col_func, dir.to_col_fn)
+    from_node = Node(
+        function=df_from[dir.from_col_func].iloc[0],
+        filename=df_from[dir.from_col_fn].iloc[0]
+    )
+    to_nodes = []
+    for _, row in df_to.iterrows():
+        to_nodes.append(
+            Node(
+                function=row[dir.to_col_func], filename=row[dir.to_col_fn]
+            )
+        )
+    chains_df = find_all_chains(df, from_node, to_nodes, dir)
+    return chains_df
+
+
+################################################################################
+
+
+def getargs():
+    desc = "Find function call chains reachable from the input function."\
+        " The end of call chain is defined with second argument which is a"\
+        " regular expression matching one or more function names. The output"\
+        " is either the database of detected call chains in csv format or image"\
+        " of the detected subgraph"
+
+    epil = "Example: ./%s --from_function sock_recvmsg --to_function '^x64_sys'"\
+        "--calls calls.csv" % os.path.basename(__file__)
+    parser = argparse.ArgumentParser(description=desc, epilog=epil)
+
+    required_named = parser.add_argument_group('required named arguments')
+    help = "function call database csv file"
+    required_named.add_argument("--calls", help=help, required=True)
+    help = "function name from where the search begins (literal match)"
+    required_named.add_argument("--from_function", help=help, required=True)
+    help = "function name where call chain ends (regex match)"
+    required_named.add_argument("--to_function", help=help, required=True)
+
+    help = "name of the output file containing detected chains"
+    parser.add_argument("--out", help=help, default="chains.csv")
+    choices = ["left", "right", "both"]
+    help = "selects search direction."
+    parser.add_argument("--direction", help=help, choices=choices, default="right")
+    help = "select cutoff lenght for path search"
```

Comment:
cutoff *length*

Also, maybe the script output should somehow indicate - at least when the user does not specify a value for the cutoff argument - that the output results were generated with default cutoff of 10 and that there might actually be more matches if a larger cutoff was given

Reply:
Added warning info when cutoff is activated. Moved default to None.